### DISCUSSION TITLE
Can `std::search` be used to search for a key?

### DISCUSSION THREAD

**Discussion by mlund** [nlohmann/json#2703] [Q&A] [answered]
I'm trying to use [`std::search`](https://en.cppreference.com/w/cpp/algorithm/search) to look for a range of keys:

~~~ cpp
std::vector<std::string> valid_keys = {"a", "b", "c"};
nlohmann::json j = {{"b", 1.0}};
auto compare = [](const auto& item, const auto& valid_key) { return item.key() == valid_key; };
auto it = std::search(j.items().begin(), j.items().end(), valid_keys.begin(), valid_keys.end(), compare);
~~~
This however is a compile error:
~~~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:1503:12: error: no matching function for call to '__search'
    return _VSTD::__search<typename add_lvalue_reference<_BinaryPredicate>::type>
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__config:840:15: note: expanded from macro '_VSTD'
#define _VSTD std::_LIBCPP_ABI_NAMESPACE

/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/functional:3040:1: note: candidate function template not viable: no known conversion from 'typename iterator_traits<iteration_proxy_value<iter_impl<const basic_json<> > > >::iterator_category' (aka 'std::__1::input_iterator_tag') to 'std::__1::forward_iterator_tag' for 6th argument
__search(_ForwardIterator1 __first1, _ForwardIterator1 __last1,
~~~

Any suggestions how to use `std::search` to search for a key, as opposed to value? My immediate thought was to use `items()`, but as shown doesn't work.

_Background:_ The above example is simplified and in my real application, the `allowed_keys` is a `std::map` where values are to perform a conditional operation, depending on the key.

**Comment by nlohmann** [comment, marked as answer]
`std::search` requires the iterators to have the forward iterator trait. For reasons I currently can't remember myself, the iterators returned by `items()` are only input iterators. Changing `nlohmann::detail::iteration_proxy_value::iterator_category` to `std::forward_iterator_tag` lets your code compile.

**Reply by nlohmann** [reply]
(I think the reason I only expected iteration to be valid once is the generation of keys from array indices. I think some additional work is needed to change this to a proper forward iterator. Feel free to create an issue for this, though I cannot promise to fix this any time soon.)
